Use rvalue constructor when building a concatenation_exprt

tautschnig opened the pull request:
Also don't do any unnecessary type cast or nesting of expressions.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Use rvalue constructor when building a concatenation_exprt

Also don't do any unnecessary type cast or nesting of expressions.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b4eedf587b3a2036b32db5b8f4875ac280de9b68).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95121619

kroening commented on src/solvers/smt2/smt2_parser.cpp:
```diff
@@ -438,11 +438,8 @@ exprt smt2_parsert::function_application_fp(const exprt::operandst &op)
-    typecast_exprt(c, ieee_float_spect(width_f, width_e).to_type()));
+  return concatenation_exprt(
+    exprt::operandst(op), ieee_float_spect(width_f, width_e).to_type());
 }
```
There's an implicit type conversion now happening here, as part of the conversion.
That will work with our flattening-backend, but will fail with the SMT2 backend, where these conversions need to be casts.

tautschnig replied:
Ok, good to know, I'll fix this - but I was guided by other places doing the same, so I will actually fix it in multiple places.

tautschnig replied:
I have added a second commit that is the result of reviewing all uses of `concatenation_exprt`.

kroening replied:
Apologies, I should have been clearer. Unfortunately, the meaning of typecast_exprt is complicated. It does a semantic cast, i.e., it attempts to preserve the _interpretation_ of the expression. The current code constructs an unisignedbv_typet expression, which is then casted to floatbv. This means that say 0u becomes 0.0f; what is needed here is a reinterpretation, i.e., a cast that doesn't change the representation.

In an ideal world, I think these two casts should be two separate expressions, which would make it clear whether reinterpretation happens or an actual transformation. In our current set-up, the way to do the reinterpretation is to first produce our bv_typet(), which has no meaning, and then cast that to the floating-point type.

tautschnig replied:
I have pushed another update, now using `bv_typet`. Is that now correct? Note that I have made this change in two places.

tautschnig committed:
Review types of concatenation_exprt

The type should be an unsignedbv_typet, and any other type requires an explicit
cast to make sure the SMT back-end can correctly handle it.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 332ab5f3a2f8b3b5a44ce19549f4628898505514).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95161228

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5de3f40c00d9d364613521cb608582df6a70d9f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95248163

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7c5a8760f4c53cab600ce6edf8b86d8a318ab31d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95391375
